hera_cal/nucal.py

```diff
@@ -1013,3 +1032,516 @@ def project_u_model_comps_on_spec_axis(u_model_comps, spectral_filters):
+    phase = jnp.einsum('bn,ntf->btf', idealized_blvecs, model_parameters["tip_tilt"])
+
+    # Compute model from foreground estimates and amplitude
+    model_r = (model_parameters["amplitude"]) * (fg_model_r * jnp.cos(phase) - fg_model_i * jnp.sin(phase))
+    model_i = (model_parameters["amplitude"]) * (fg_model_i * jnp.cos(phase) + fg_model_r * jnp.sin(phase))
+    
+    # Compute loss using weights and foreground model
+    return jnp.mean((jnp.square(model_r - data_r) + jnp.square(model_i - data_i)) * wgts)
+
+@jax.jit
+def _calibration_loss_function(model_parameters, data_r, data_i, wgts, spectral_filters, spatial_filters, idealized_blvecs):
+    """
+    Function which computes the value of the loss from the degenerate parameters, DPSS foreground components, and the data 
+    
+    Parameters:
+    ----------
+    model_parameters : dictionary
+        Parameters for fitting
+    data_r : np.ndarray
+        Array of real component of data with shape (Ntimes, Nbls)
+    data_i : np.ndarray
+        Array of imaginary component of data with shape (Ntimes, Nbls)
+    wgts : np.ndarray
+        Array of weights with shape (Ntimes, Nbls)
+    spectral_filters : np.ndarray
+        Array of spectral filters with shape (Nfreqs, Nfilters)
+    spatial_filters : List
+        List of spatial filters for each baseline in the group
+    idealized_blvecs : np.ndarray
+        Array of idealized baseline vectors with shape (Nbls, Ndims)
+
+    Returns:
+    -------
+    loss : float
+        Mean squared error between data and foreground model
+    """
+    # Compute foreground model from the model_parameters and DPSS filters
+    fg_model_r, fg_model_i = _foreground_model(model_parameters, spectral_filters, spatial_filters)
+
+    # Compute loss
+    return _mean_squared_error(model_parameters, data_r, data_i, wgts, fg_model_r, fg_model_i, idealized_blvecs)
+
+def _nucal_post_redcal(
+        data_r, data_i, wgts, model_parameters, optimizer, spectral_filters, spatial_filters, idealized_blvecs,
+        maxiter=100, convergence_criteria=1e-10, minor_cycle_maxiter=10
+    ):
+    """
+    Function to perform frequency redundant calibration using gradient descent. Calibrates the 
+    data by fitting a foreground model to the data using DPSS filters and estimating the redundant
+    calibration degeneracies that lead to the smoothest calibrated visibilities. Intended to only
+    be only after performing redundant calibration with redcal.RedundantCalibrator.
+
+    Parameters:
+    ----------
+    data_r : np.ndarray
+        Array of real component of data with shape (Ntimes, Nbls)
+    data_i : np.ndarray
+        Array of imaginary component of data with shape (Ntimes, Nbls)
+    wgts : np.ndarray
+        Array of weights with shape (Ntimes, Nbls)
+    model_parameters : dictionary
+        Parameters for fitting
+    optimizer : optax optimizer
+        Optimizer to use for gradient descent
```

Comment:
Might be useful to also allow keying this on a string that's in `nucal.OPTIMIZERS`